wpilibc/src/main/native/include/frc/LEDPattern.h

```diff
@@ -0,0 +1,325 @@
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <functional>
+#include <span>
+#include <utility>
+
+#include <units/length.h>
+#include <units/time.h>
+#include <units/velocity.h>
+
+#include "frc/AddressableLED.h"
+#include "util/Color.h"
+#include "util/Color8Bit.h"
+
+namespace frc {
+
+/**
+ * Sets the LED at the given index to the given color.
+ */
+using LEDWriterFn = std::function<void(int, frc::Color)>;
+
+/**
+ * Accepts a data buffer (1st argument) and a callback (2nd argument) for
+ * writing data.
+ */
+using LEDPatternFn =
+    std::function<void(std::span<frc::AddressableLED::LEDData>, LEDWriterFn)>;
+
+class LEDPattern {
+ public:
+  explicit LEDPattern(const LEDPatternFn& impl);
+
+  ~LEDPattern() = default;
+
+  /**
+   * Writes the pattern to an LED buffer. Dynamic animations should be called
+   * periodically (such as with a command or with a periodic method) to refresh
+   * the buffer over time.
+   *
+   * This method is intentionally designed to use separate objects for reading
+   * and writing data. By splitting them up, we can easily modify the behavior
+   * of some base pattern to make it scroll, blink, or breathe by intercepting
+   * the data writes to transform their behavior to whatever we like.
+   */
+  void ApplyTo(std::span<frc::AddressableLED::LEDData> data,
+               LEDWriterFn writer);
+  void ApplyTo(std::span<frc::AddressableLED::LEDData> data);
+
+  /**
+   * Creates a pattern that displays this one in reverse. Scrolling patterns
+   * will scroll in the opposite direction (but at the same speed). It will
+   * treat the end of an LED strip as the start, and the start of the strip as
+   * the end. This can be useful for making ping-pong patterns that travel from
+   * one end of an LED strip to the other, then reverse direction and move back
+   * to the start. This can also be useful when working with LED strips
+   * connected in a serpentine pattern (where the start of one strip is
+   * connected to the end of the previous one).
+   *
+   * @return the reverse pattern
+   */
+  [[nodiscard]]
+  LEDPattern Reversed();
+
+  /**
+   * Creates a pattern that displays this one, but offset by a certain number of
+   * LEDs. The offset pattern will wrap around, if necessary.
+   *
+   * @param offset how many LEDs to offset by
+   * @return the offset pattern
+   */
+  [[nodiscard]]
+  LEDPattern OffsetBy(int offset);
+
+  /**
+   * Creates a pattern that plays this one scrolling up the buffer. The velocity
+   * controls how fast the pattern returns back to its original position, and is
+   * in terms of the length of the LED strip; scrolling across a segment that is
+   * 10 LEDs long will travel twice as fast as on a segment that's only 5 LEDs
+   * long (assuming equal LED density on both segments).
+   */
+  [[nodiscard]]
+  LEDPattern ScrollAtRelativeSpeed(double velocity);
```

Comment:
Would `units::hertz_t` be a logical type for `velocity`? (Optionally renaming the parameter)

Reply:
If it's a patterns/sec thing, then yea.